pkg/gatewayconfigurationserver/gatewayconfigurationserver.go

```diff
@@ -0,0 +1,155 @@
+	RequireAuth bool `name:"require-auth" description:"Require authentication for the HTTP endpoints."`
+}
+
+// GatewayConfigurationServer implements the Gateway Configuration Server component.
+type GatewayConfigurationServer struct {
+	*component.Component
+	config *Config
+
+	registry ttnpb.GatewayRegistryClient
+
+	ctx context.Context
+}
+
+// RegisterRoutes registers the web frontend routes.
+func (gcs *GatewayConfigurationServer) RegisterRoutes(server *ttnweb.Server) {
+	middleware := []echo.MiddlewareFunc{
+		gcs.handleError(),
+		gcs.validateAndFillIDs(),
+	}
+	if gcs.config.RequireAuth {
+		middleware = append(middleware, gcs.requireGatewayRights(ttnpb.RIGHT_GATEWAY_INFO))
+	}
+	group := server.Group(ttnpb.HTTPAPIPrefix+"/gcs/gateways/:gateway_id", middleware...)
+	group.GET("/global_conf.json", func(c echo.Context) error {
+		return gcs.handleGet(c)
+	})
+}
+
+// New returns new *GatewayConfigurationServer.
+func New(c *component.Component, conf *Config, opts ...Option) (*GatewayConfigurationServer, error) {
+	gcs := &GatewayConfigurationServer{
+		Component: c,
+		config:    conf,
+		ctx:       c.FillContext(c.Context()),
+	}
+	for _, opt := range opts {
+		opt(gcs)
+	}
+
+	cups := conf.BasicStation.NewServer(c)
+	_ = cups
+
+	c.RegisterWeb(gcs)
+	return gcs, nil
+}
+
+// Option represents a Gateway Configuration Server option.
+type Option func(gcs *GatewayConfigurationServer)
+
+// WithRegistry sets the gateway registry for the given server.
+func WithRegistry(registry ttnpb.GatewayRegistryClient) Option {
+	return func(gcs *GatewayConfigurationServer) {
+		gcs.registry = registry
+	}
+}
+
+// WithContext sets the context for the given server.
+func WithContext(ctx context.Context) Option {
+	return func(gcs *GatewayConfigurationServer) {
+		gcs.ctx = ctx
+	}
+}
+
+func (gcs *GatewayConfigurationServer) handleGet(c echo.Context) error {
+	ctx := gcs.getContext(c)
+	gtwID := c.Get(gatewayIDKey).(ttnpb.GatewayIdentifiers)
+	gtw, err := gcs.getGateway(ctx, gtwID, types.FieldMask{
+		Paths: []string{
+			"gateway_server_address",
+			"frequency_plan_id",
+		},
+	})
+	if err != nil {
+		return err
+	}
+	config, err := semtechudp.Build(gtw, gcs.FrequencyPlans)
+	if err != nil {
+		return err
+	}
+	return c.JSONPretty(http.StatusOK, config, "\t")
+}
+
+func (gcs *GatewayConfigurationServer) getContext(c echo.Context) context.Context {
+	ctx := c.Request().Context()
+	ctx = gcs.FillContext(ctx)
+	md := metadata.New(map[string]string{
+		"authorization": c.Request().Header.Get(echo.HeaderAuthorization),
+	})
+	if ctxMd, ok := metadata.FromIncomingContext(ctx); ok {
+		md = metadata.Join(ctxMd, md)
+	}
+	return metadata.NewIncomingContext(ctx, md)
+}
+
+var (
+	errEmptyIdentifiers = errors.Define("empty_identifiers", "empty identifiers")
+)
+
+func (gcs *GatewayConfigurationServer) getGateway(ctx context.Context, ids ttnpb.GatewayIdentifiers, fieldmask types.FieldMask) (*ttnpb.Gateway, error) {
+	ctx = gcs.FillContext(ctx)
```

Comment:
`getGateway` is only called by `handleGet` which already called `getContext` that calls `FillContext`.